File: ttsim/front/ttdecode/tests/python/test_decode_all_rv32.py
import os
import pytest

try:
    from ttdecode.core import _core as core
except Exception:
    core = None

pytestmark = pytest.mark.skipif(core is None, reason="_core module not available; build with Python bindings enabled")

def test_all_rv32_mnemonics_match():
    path = os.path.join(os.path.dirname(__file__), "../../../../config/llk/instruction_sets/rv32/assembly.yaml")
    if not os.path.exists(path):
        pytest.skip("rv32 assembly.yaml not found")

    iset = core.isa.get_instruction_set(path, core.isa.instruction_kind.rv32)

    def set_bits(word, start, size, value):
        if size == 32:
            mask = 0xFFFFFFFF
        else:
            mask = (1 << size) - 1
        mask <<= start
        word &= ~mask
        word |= ((value & (0xFFFFFFFF if size == 32 else ((1 << size) - 1))) << start)
        return word

    for mnemonic, ins in iset.items():
        if mnemonic in ["SLLI.R1", "SRLI.R1", "SRAI.R1"]:
            continue
        word = 0
        word = set_bits(word, 0, 7, int(ins.opcode))
        for enc in ins.encodings.values():
            word = set_bits(word, int(enc.start), int(enc.size), int(enc.value))

        di = core.decode.rv32_decode(word, path, True)
        assert di.mnemonic == mnemonic
        assert di.opcode == int(ins.opcode)

File: ttsim/front/ttdecode/tests/cpp/test_decode_all_rv32.cpp
#include "decode/rv32.hpp"
#include "isa/isa.hpp"
#include <gtest/gtest.h>
#include <string>
#include <map>

static std::string get_assembly_path(const std::string &rel) {
    std::string current_file(__FILE__);
    size_t pos = current_file.find_last_of("/\\");
    std::string dir = (pos != std::string::npos) ? current_file.substr(0, pos) : ".";
    return dir + "/" + rel;
}

static std::uint32_t set_bits(std::uint32_t word, std::uint32_t start, std::uint32_t size, std::uint32_t value) {
    const std::uint32_t mask = ((size == 32 ? 0xFFFFFFFFu : ((1u << size) - 1u)) << start);
    word &= ~mask;
    word |= ((value & (size == 32 ? 0xFFFFFFFFu : ((1u << size) - 1u))) << start);
    return word;
}

TEST(DecodeAll, RV32_AllInstructionsMatchMnemonic) {
    const std::string file = get_assembly_path("../../../../config/llk/instruction_sets/rv32/assembly.yaml");
    auto iset = ttdecode::isa::get_instruction_set(file, ttdecode::isa::instruction_kind::rv32);

    for (const auto &kv : iset) {
        const std::string &mnemonic = kv.first;
        const ttdecode::isa::instruction &ins = kv.second;

        // Skip RV64-specific instructions that should not be decoded by RV32 decoder
        if (mnemonic == "SLLI.R1" || mnemonic == "SRLI.R1" || mnemonic == "SRAI.R1") {
            continue;
        }

        std::uint32_t word = 0u;
        word = set_bits(word, 0u, 7u, static_cast<std::uint32_t>(ins.opcode));
        for (const std::pair<const std::uint8_t, ttdecode::isa::encoding> &e_kv : ins.encodings) {
            const ttdecode::isa::encoding &enc = e_kv.second;
            word = set_bits(word, enc.start, enc.size, enc.value);
        }

        ttdecode::decode::decoded_instruction di = ttdecode::decode::rv32_decode(word, iset, true);
        ASSERT_TRUE(di.mnemonic.has_value());
        EXPECT_EQ(*di.mnemonic, mnemonic);
        ASSERT_TRUE(di.opcode.has_value());
        EXPECT_EQ(*di.opcode, static_cast<std::uint32_t>(ins.opcode));
    }
}
